numpy/core/src/multiarray/dtype_traversal.c

```diff
@@ -0,0 +1,455 @@
+        --size;
+    }
+    return 0;
+}
+
+
+NPY_NO_EXPORT int
+npy_get_clear_object_strided_loop(
+        void *NPY_UNUSED(traverse_context), PyArray_Descr *NPY_UNUSED(descr),
+        int NPY_UNUSED(aligned), npy_intp NPY_UNUSED(fixed_stride),
+        traverse_loop_function **out_loop, NpyAuxData **out_auxdata,
+        NPY_ARRAYMETHOD_FLAGS *flags)
+{
+    *flags = NPY_METH_REQUIRES_PYAPI|NPY_METH_NO_FLOATINGPOINT_ERRORS;
+    *out_loop = &clear_object_strided_loop;
+    return 0;
+}
+
+
+/**************** Structured DType clear funcationality ***************/
+
+/*
+ * Note that legacy user dtypes also make use of this.  Someone managed to
+ * hack objects into them by adding a field that contains objects and this
+ * remains (somewhat) valid.
+ * (Unlike our voids, those fields must be hardcoded probably, but...)
+ *
+ * The below functionality mirrors the casting functionality relatively
+ * closely.
+ */
+
+typedef struct {
+    npy_intp src_offset;
+    NPY_traverse_info info;
+} single_field_clear_data;
+
+typedef struct {
+    NpyAuxData base;
+    npy_intp field_count;
+    single_field_clear_data fields[];
+} fields_clear_data;
+
+
+/* transfer data free function */
+static void
+fields_clear_data_free(NpyAuxData *data)
+{
+    fields_clear_data *d = (fields_clear_data *)data;
+
+    for (npy_intp i = 0; i < d->field_count; ++i) {
+        NPY_traverse_info_xfree(&d->fields[i].info);
+    }
+    PyMem_Free(d);
+}
+
+
+/* transfer data copy function */
+static NpyAuxData *
+fields_clear_data_clone(NpyAuxData *data)
+{
+    fields_clear_data *d = (fields_clear_data *)data;
+
+    npy_intp field_count = d->field_count;
+    npy_intp structsize = sizeof(fields_clear_data) +
+                    field_count * sizeof(single_field_clear_data);
+
+    /* Allocate the data and populate it */
+    fields_clear_data *newdata = PyMem_Malloc(structsize);
+    if (newdata == NULL) {
+        return NULL;
+    }
+    newdata->base = d->base;
+    newdata->field_count = 0;
+
+    /* Copy all the fields transfer data */
+    single_field_clear_data *in_field = d->fields;
+    single_field_clear_data *new_field = newdata->fields;
+
+    for (; newdata->field_count < field_count;
+                newdata->field_count++, in_field++, new_field++) {
+        new_field->src_offset = in_field->src_offset;
+
+        if (NPY_traverse_info_copy(&new_field->info, &in_field->info) < 0) {
+            fields_clear_data_free((NpyAuxData *)newdata);
+            return NULL;
+        }
+    }
+
+    return (NpyAuxData *)newdata;
+}
+
+
+static int
+traverse_fields_function(
+        void *traverse_context, PyArray_Descr *NPY_UNUSED(descr),
+        char *data, npy_intp N, npy_intp stride,
+        NpyAuxData *auxdata)
+{
+    fields_clear_data *d = (fields_clear_data *)auxdata;
+    npy_intp i, field_count = d->field_count;
+
+    /* Do the traversing a block at a time for better memory caching */
+    const npy_intp blocksize = NPY_LOWLEVEL_BUFFER_BLOCKSIZE;
```

Comment:
Are you sure this is more performant than a simpler approach?

Reply:
This is the approach we use everywhere.  But yes, it should make sense for cache friendliness.